### PR TITLE
Metric labels do not include hostname when two queries are used

### PR DESCRIPTION
**Pull request by lpetrora**
This change adds a custom label input box and allows the user to customize the labels. It corresponding filter is created, these variables are available:
* $filter_site: Site name
* $filter_host_name: Host name
* $filter_host_in_group: Group containing the host
* $filter_service: Service name
* $filter_service_in_group: Group containing the service

$label is always available and contains the original label sent by Checmk.

CMK-15138

### CODE REVIEW

**Review by BenediktSeidl — CHANGES_REQUESTED** [review]
changing the label triggers a api call to checkmk, although it is not really needed as we already have all the data we need. can you give it a try to see if you can accomplish the updating of the lables without triggering a new api request to checkmk?
i fear we would have to change quite a lot of the current architecture, but maybe it's doable?

File: src/ui/QueryEditor.tsx [hunk 1]
```diff
@@ -109,6 +112,35 @@ export const QueryEditor = (props: Props): JSX.Element => {
...
 
+  const labelField = (
+    <GenericField
+      requestSpecKey="label"
+      label={labelForRequestSpecKey('label', requestSpec)}
+      value={qLabel}
+      onChange={setQLabel}
+      dataTestId="custom-label-field"
+    >
+      <Alert title="Variables on custom label" severity="info">
```
[the remark this answers is not in the document]
Comment (lpetrora) [reply src/ui/QueryEditor.tsx:123]: I added a toggle button witch displays a popup

image

File: src/backend/rest.ts [hunk 2]
```diff
@@ -265,9 +267,6 @@ export default class RestApiBackend implements Backend {
       throw 'Query is missing required fields';
     }
 
-    range.from.set('milliseconds', 0);
```
[the remark this answers is not in the document]
Comment (lpetrora) [reply src/backend/rest.ts:268]: Yes, it was an accident

File: src/ui/components.tsx [hunk 3]
```diff
@@ -288,6 +288,51 @@ export const Filter = <T extends RequestSpecNegatableOptionKeys>(props: FilterPr
...
+  tooltip?: string;
+  dataTestId?: string;
+}
+export const GenericField = <T extends RequestSpecStringKeys>(props: GenericFieldProps<T>) => {
+  const { label, width = 32, onChange, value, requestSpecKey, children = null, tooltip } = props;
+  const { dataTestId = `${requestSpecKey}-filter-input` } = props;
+
+  const [textValue, setTextValue] = React.useState(value !== undefined ? value : '');
+
+  const debouncedOnChange = React.useMemo(
```
[the remark this answers is not in the document]
Comment (lpetrora) [reply src/ui/components.tsx:304]: Yes, the idea is to refactor Filter to use GenericField and simplify the code

File: src/utils.ts [hunk 4]
```diff
@@ -276,3 +277,23 @@ export function toLiveStatusQuery(filter: Partial<FiltersRequestSpec>, table: 'h
...
+
+export function updateMetricTitles(metrics: GrapResponse[], query: CmkQuery) {
+  const createRegexp = (value: string) => new RegExp(value.replace(/[-\\^$*+?.()|[\]{}]/g, '\\$&'), 'g');
+
+  const titleTemplate = query.requestSpec?.label || label.ORIGINAL;
+
+  if (titleTemplate !== label.ORIGINAL) {
+    metrics.forEach((metric) => {
+      metric.title = titleTemplate
+        .replace(createRegexp(label.ORIGINAL), metric.title)
```
[the remark this answers is not in the document]
Comment (lpetrora) [reply src/utils.ts]: I modified it and now we use  getTemplateSrv. It allows us to use query parameters (filters) and Grafana's variables for labels names

**Review by BenediktSeidl — COMMENTED** [review]
regarding the "change label without backend request":

this is where the request is triggered:
https://github.com/Checkmk/grafana-checkmk-datasource/blob/8c2a87a5b020f95056ea57f84de7a3daa94653be/src/ui/QueryEditor.tsx#L48-L55

this is the new url to the docs, the old one no longer works:
https://react.dev/reference/react/useState#storing-information-from-previous-renders

i thought: we could check what parts of the request spec changed, and if only the label changed we call `onChange` only and if we detect that parts of the filter changed, we  call `onRunQuery`.

maybe we should also think about a better idea than this JSON.stringify thing, but this is a quest for another pr.

**Comment by BenediktSeidl** [comment]
> regarding the "change label without backend request":

as you already mentioned my last idea does not work. i looked at the code again, and maybe my next idea could work: 
we could remember the last query, and the last query result in this function:

https://github.com/Checkmk/grafana-checkmk-datasource/blob/main/src/DataSource.ts#L30-L35

and if they current query is the same as the last query (minus the label change of course) we could use the old result and just interpolate the changed label.

but this feels a bit hacky. probably it does not work at all, because the time range is relative, so we will get a changed timerange with the label change, so the queries differ...

maybe the whole idea of saving one http request is a premature optimization? we can skip it for now, see how everything works out and come back to this idea if we see that its really necessary.

sorry about the confusion.

**Review by BenediktSeidl — APPROVED** [review]

**Pull request merged**